### PR TITLE
Support slot order

### PR DESCRIPTION
**Pull request by luckyxiaoqiang**
### Describe what this PR does / why we need it
Support slot order. 

Slots in a slot chain are sorted by order value of slots in increasing order.  Default slot chain has a fixed default order and have no relation to the add sequence. User extended slot can set the order to make it be in front of or at the back of a sentinel build-in slot easily. 

### Does this pull request fix one issue?
#317 

### Special notes for reviews

### CODE REVIEW

File: core/base/slot_chain.go [hunk 1]
```diff
@@ -125,15 +129,16 @@ func (sc *SlotChain) RangeStatPrepareSlot(f func(slot StatPrepareSlot)) {
...
 
-func (sc *SlotChain) AddStatPrepareSlotFirst(s StatPrepareSlot) {
-	ns := make([]StatPrepareSlot, 0, len(sc.statPres)+1)
-	// add to first
-	ns = append(ns, s)
-	sc.statPres = append(ns, sc.statPres...)
-}
-
-func (sc *SlotChain) AddStatPrepareSlotLast(s StatPrepareSlot) {
+func (sc *SlotChain) InsertStatPrepareSlotByOrder(s StatPrepareSlot) {
```
Comment (louyuting) [inline core/base/slot_chain.go:132]: Maybe suffix “ByOrder” is not needed?
such as "AddStatPrepareSlot()" is ok?

File: core/base/slot_chain.go [hunk 2]
```diff
@@ -125,15 +129,16 @@ func (sc *SlotChain) RangeStatPrepareSlot(f func(slot StatPrepareSlot)) {
...
 	sc.statPres = append(sc.statPres, s)
+	i := len(sc.statPres) - 1
```
Comment (louyuting) [inline core/base/slot_chain.go:134]: Can the standard pkg "sort" support sorting logic?
such as ‘sort.Slice()’？

Comment (luckyxiaoqiang) [reply core/base/slot_chain.go]: Refined.

File: core/base/slot_chain.go (hunk 1 above)
Comment (luckyxiaoqiang) [reply core/base/slot_chain.go]: Refined.

File: core/base/slot_chain.go [hunk 3]
```diff
@@ -157,14 +158,11 @@ func (sc *SlotChain) RangeRuleCheckSlot(f func(slot RuleCheckSlot)) {
...
-func (sc *SlotChain) AddRuleCheckSlotFirst(s RuleCheckSlot) {
-	ns := make([]RuleCheckSlot, 0, len(sc.ruleChecks)+1)
-	ns = append(ns, s)
-	sc.ruleChecks = append(ns, sc.ruleChecks...)
-}
-
-func (sc *SlotChain) AddRuleCheckSlotLast(s RuleCheckSlot) {
+func (sc *SlotChain) AddRuleCheckSlot(s RuleCheckSlot) {
 	sc.ruleChecks = append(sc.ruleChecks, s)
+	sort.Slice(sc.ruleChecks, func(i, j int) bool {
```
Comment (louyuting) [inline core/base/slot_chain.go:163]: Maybe we need unit test to cover rank result.

File: core/base/slot_chain.go [hunk 4]
```diff
@@ -1,22 +1,27 @@
...
+	// Order returns the order value of slot.
+	// Slots will be sorted according to order value in increasing order.
+	Order() uint32
 }
 
 // StatPrepareSlot is responsible for some preparation before statistic
 // For example: init structure and so on
 type StatPrepareSlot interface {
-	NamedSlot
+	BaseSlot
```
Comment (louyuting) [inline core/base/slot_chain.go:24]: Maybe we need some comments to describe the order logic.
StatPrepareSlot、RuleCheckSlot、StatSlot is the independent bucket, the ranking logic is independent.

File: core/base/slot_chain.go [hunk 5]
```diff
@@ -28,7 +33,7 @@ type StatPrepareSlot interface {
 // RuleCheckSlot is rule based checking strategy
 // All checking rule must implement this interface.
 type RuleCheckSlot interface {
-	NamedSlot
+	BaseSlot
```
Comment (louyuting) [inline core/base/slot_chain.go:36]: Same as upon

File: core/base/slot_chain.go [hunk 6]
```diff
@@ -39,7 +44,7 @@ type RuleCheckSlot interface {
 // StatSlot is responsible for counting all custom biz metrics.
 // StatSlot would not handle any panic, and pass up all panic to slot chain
 type StatSlot interface {
-	NamedSlot
+	BaseSlot
```
Comment (louyuting) [inline core/base/slot_chain.go:47]: Same with upon

**Comment by louyuting** [comment]
Some suggestion about default order of embedded  slot.
1. In each bucket(PrepareStatSlot Bucket、CheckSlot Bucket、StatSlot Bucket), The default order should from local to external。Such as CheckSlot, it should be system、flow、isolation、hotspot、circuitbreaker;
2. The StatSlot order should keep consistent with CheckSlot bucket。 Such as: flowStat、hotspotStat、circuitbreakerStat.

File: core/base/slot_chain.go (hunk 3 above)
Comment (luckyxiaoqiang) [reply core/base/slot_chain.go:163]: There are already related unit tests like `TestSlotChain_AddStatPrepareSlot`. Is is sufficient?

Comment (louyuting) [reply core/base/slot_chain.go]: Okay for me.

**Comment by luckyxiaoqiang** [comment]
> Some suggestion about default order of embedded slot.
> 
> 1. In each bucket(PrepareStatSlot Bucket、CheckSlot Bucket、StatSlot Bucket), The default order should from local to external。Such as CheckSlot, it should be system、flow、isolation、hotspot、circuitbreaker;
> 2. The StatSlot order should keep consistent with CheckSlot bucket。 Such as: flowStat、hotspotStat、circuitbreakerStat.

Agree. Refined.

**Comment by louyuting** [comment]
@liqiangz @sanxun0325  Could you please take a look on this PR ?

File: core/circuitbreaker/slot.go [hunk 7]
```diff
@@ -5,7 +5,8 @@ import (
 )
 
 const (
-	RuleCheckSlotName = "sentinel-core-circuit-breaker-rule-check-slot"
+	RuleCheckSlotName  = "sentinel-core-circuit-breaker-rule-check-slot"
+	RuleCheckSlotOrder = 5000
```
Comment (sanxun0325) [inline core/circuitbreaker/slot.go:9]: It is better to define the order of slots in a uniform location, which may seem more intuitive。

File: core/stat/stat_prepare_slot.go [hunk 8]
```diff
@@ -5,7 +5,8 @@ import (
 )
 
 const (
-	PrepareSlotName = "sentinel-core-stat-resource-node-prepare-slot"
+	PrepareSlotName  = "sentinel-core-stat-resource-node-prepare-slot"
+	PrepareSlotOrder = 1000
```
Comment (sanxun0325) [inline core/stat/stat_prepare_slot.go:9]: Does the definition of a slot order, starting at 1000, have any particular meaning?

File: core/circuitbreaker/slot.go (hunk 7 above)
Comment (luckyxiaoqiang) [reply core/circuitbreaker/slot.go:9]: My first version is like what you recommended, current version is the result of discussion with @louyuting .

File: core/stat/stat_prepare_slot.go (hunk 8 above)
Comment (luckyxiaoqiang) [reply core/stat/stat_prepare_slot.go:9]: Leave a gap for the user to add extended slots before it.

Comment (sanxun0325) [reply core/stat/stat_prepare_slot.go:9]: > Leave a gap for the user to add extended slots before it.

Yes,The main thing I'm trying to say is the order of slots 1000, why isn't it 2000, or 10000.There should be a specification for it.  I don't know if there are any other better ideas for defining the slot order that Sentinel provides by default。

**Review by louyuting — APPROVED** [review]

**Review by sczyh30 — APPROVED** [review]
LGTM

**Pull request merged**